Initial documentation for the envelope encryption filter (#932)
kroxylicious/kroxylicious: docs/available-filters/envelope-encryption/envelope-encryption.adoc

:kms-api-javadoc: https://javadoc.io/doc/io.kroxylicious/kroxylicious-kms/latest
:encryption-api-javadoc: https://javadoc.io/doc/io.kroxylicious/kroxylicious-encryption
:design-doc: https://github.com/kroxylicious/kroxylicious/blob/main/kroxylicious-filters/kroxylicious-encryption/doc/design.adoc

= Envelope Encryption

== What is it?

A filter that provides a transparent https://kroxylicious.io/use-cases/[encryption-at-rest solution] for Apache Kafka.

At a high level, the filter works in the following way.  First, let's consider the produce side:

1. The filter intercepts produce requests sent from producing Kafka Clients.
2. The filter disassembles the produce request.
3. Each record within it is encrypted.
4. The produce request is reassembled, replacing the original records with their encrypted counterparts.
5. The filter forwards the modified produce request onto the Kafka Broker.

> 6. The broker handles the records in the normal way writing them to the transaction log.  The broker has no knowledge

Reviewer: The `transaction log` term is not used in Kafka, and might be confused with kafka's transactional log. Maybe use `data log`?

Author: I meant https://kafka.apache.org/documentation/#log. My intent was to emphasise to the reader that the broker has no knowledge of the fact that the record is encrypted.

   that the records are encrypted - to it, they are just opaque bytes.

Now, let's consider the consume side:

1. The filter intercepts the fetch responses sent by the Kafka Broker to the consuming Kafka Client.
2. The filter disassembles the fetch response.
3. Each record is decrypted.
4. The fetch response is reassembled, replacing the encrypted records with their unencrypted counterparts.
5. The filter forwards the modified fetch response onto the Kafka Client. The client has no knowledge that the record was encrypted.

The entire process is transparent from the point of view of both Kafka Client and Kafka Broker.  Neither are
aware that the records are being encrypted.  Neither client nor broker have any access to the encryption keys.

The filter encrypts the records using symmetric encryption keys.  The encryption technique employed is

> known as *Envelope Encryption*, which is technique suited for encrypting large volumes of data in a efficient manner.

Reviewer 2: ```suggestion
known as *Envelope Encryption*, which is a technique suited for encrypting large volumes of data in an efficient manner.
```

Envelope Encryption is described in the next section.

The filter integrates with a Key Management Service (KMS).  It is the Key Management Service (KMS) that has
ultimate responsibility for the safe storage of key material.  The role of the KMS is discussed later.

=== Envelope Encryption

Envelope Encryption is a technique described by https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r5.pdf[NIST
Special Publication 800-57 Part 1 Revision 5]. It is the practice of encrypting the data with a data encryption key (DEK),
then wrapping (encrypting) the DEK with a key encryption key (KEK).  In this section we discuss how Envelope Encryption is
applied by the filter to encrypt Kafka records.  For more detail, refer to the {design-doc}[design].

On the produce path, the filter is encrypting records.  The filter uses a selector to determine which KEK to use. It then
requests that the KMS generates a DEK for the KEK.  It is the DEK that is used to encrypt the record.  The original record
in the produce request is replaced by a *cipher record* which comprises the encrypted record, the encrypted DEK and some other

> data.

Reviewer 3: I understand, and like, what you are saying here but I wonder if we should be introducing a new term here, `cipher record`, rather than referring to the envelope or even the term we use in the code `parcel`? 

I think we should be using the same term in all contexts.

Reviewer: `some other data` -> `some metadata`?

On the fetch path, the filter is decrypting records.  The filter receives the *cipher record* from the Kafka Broker. The

> filter reverses the process used to  construct the *cipher record* and uses the KMS to decrypt the DEK.  The decrypted DEK is

Reviewer: used to [ ] construct the... <-- additional space

> used to decrypt the encrypted record.  The cipher record in the fetch response is then replaced with the decrypted record.

Reviewer 3: I think we should also be calling out the fact that we lean on Kafka for security here, if you can read the topic you are implicitly allowed to decrypt it.

On the produce path, the filter employs a DEK reuse strategy. This means that records sent by a single connection to

> the same topic will be encrypted using the same DEK (until a time-out or number of encryption limit is reached, whichever

Reviewer 2: something about it is hard to parse, maybe it should be hyphenated like `or a number-of-encryptions limit is reached`

occurs first).  This prevents key exhaustion and prevents excessive interactions with the KMS.

> On the consume path, the filter employs an LRU strategy to keep recently encountered DEKs decrypted in memory. This

Reviewer 3: ```suggestion
On the fetch path, the filter employs an LRU strategy to keep recently encountered DEKs decrypted in memory. This
```
For consistency with previous paragraphs.

prevents excessive interactions with the KMS.

=== Role of the KMS

The Key Management Service (KMS) is the *secure repository* for Key Encryption Keys (KEKs). The key material of the KEK
*never* leaves the confines of the KMS.  The KMS exposes cryptographic functions that operate against the KEKs stored
within it.

The KMS is also the *source* of Data Encryption Keys (DEKs).  When the KMS generates a DEK for a given KEK, it returns
the DEK (which is securely generated random data), together with the encrypted DEK (which is the same data, encrypted
using the KEK).  Note that the KMS does not *store* the DEKs - DEKs are stored encrypted as part of the Kafka record held
by the broker.

The filter uses the services of the KMS to generate DEKs, and decrypt encrypted DEKs.

The KMS must be available at runtime. If the KMS is unavailable, it will become impossible to produce or consume
records through the filter until the KMS service is restored.

The filter currently has a single KMS integration with https://www.hashicorp.com/[HashiCorp Vault&#174;].  More KMS

> integrations are planned. It is also possible for a user to implement their own KMS integration.  The implementation

Reviewer 3: `planned` implies a commitment. How about: 
```suggestion
further integrations will be driven by demand. It is also possible for a user to implement their own KMS integration (we hope where possible those can be shared).  The implementation
```

must implement the {kms-api-javadoc}/io/kroxylicious/kms/service/KmsService.html[KMS public API] and make the
implementation available on the classpath with the service loader mechanism.

> IMPORTANT: It is recommended to use a KMS in a highly available configuration.

Reviewer 3: We do need to think about what we should do with produce and fetch requests when then KMS is unavailable I'm not sure we should just reject them because they may contain records for un-encrypted topics which we should still pass through (graceful degradation)

Author: https://github.com/kroxylicious/kroxylicious/issues/970

=== Key rotation

Key rotation is the practice of periodically replacing cryptographic keys with new ones.  Using key rotation is
considered cryptographic best-practice.

The filter allows for the rotation of KEKs within the KMS. When a KEK is rotated, the new key material will be applied
to newly produced records. Existing records (which were encrypted with older versions of the KEK) remain decryptable
as long as the previous KEK version remains present in the KMS.

WARNING: If the previous KEK version is removed from the KMS, the records encrypted with that key version will become

> un-consumable (that is, the fetch will fail). In this case, the consumer offset must be advanced beyond those records.

Reviewer: Do we have a way to know which offset to seek to?

Author: No, that's a gap.  At a bare minimum we should have a script.

=== What exactly gets encrypted?

The filter currently encrypts only *record values*.  Record keys, headers, timestamps are not encrypted.

> Null record values produced are passed through unencrypted.  This serves to allow tombstoning.

Reviewer 3: I know what your referring too but I think it assumes too much knowledge of Kafka internals for our intended audience. I think we should either explain what tombstones are and how Kafka uses and represents them while explaining that we still support them or just ignore it. As a user I would just expect the filter to support compacted topics unless it explicitly said it didn't do so.

Thoughts?

Reviewer: Maybe just adding a reference link to kafka doc: https://kafka.apache.org/documentation/#design_compactionbasics

Reviewer: Could you explain where we address this comment?

The filter supports use-cases where some Kafka topics are configured for encryption while others are left to be

> unencrypted.  Transactions producing to both encrypted and unencrypted topics are unsupported?

Reviewer 3: I admit that I haven't given it much thought, but why would transactions be affected by encryption? The transaction semantics are enforced by the broker and as far as it is concerned messages are just opaque byte arrays anyway, why does the presence or absence or a mix affect that handling?

Reviewer: transaction records include control record, which is a record without content (ex: abort, commit...). I remember @robobario was working on this before?

Reviewer 2: Yes, we have already implemented the work to preserve control batches (#910) and ensure the metadata for each batch within a MemoryRecords is preserved. I think this means that transactions will work across encrypted and unencrypted topics. We could extend the transactional integration test to check it.

Reviewer 3: We should definitely extend the test coverage

== How to use the filter

There are three steps to using the filter.

> 1. Setting up the KMS.

Reviewer 3: Should we assume that people deploying the filter need to setup a KMS instance? I'd assume they already have one. Though I think giving instructions for setting one up for testing etc are valuable I think it should be an aside rather than step one.

Author: My starting point was that the KMS exists already.  The intent of this section was to tell the user what features etc need to be enabled (enabling transit engine, for instance) or what information needs to be extracted (a transit engine url) in order to start the integration.

2. Configuring the filter within Kroxylicious.
3. Establishing the encryption key(s) within the KMS that will be used to encrypt the topics.

These steps are described in the next sections.

=== Setting up the KMS

In order to set up the KMS provider ready to use with the filter, follow these KMS provider specific steps.

include::hashicorp-vault/setup.adoc[leveloffset=3]

=== Filter Configuration

The filter is configured as part of the filter chain in the following way:

[source, yaml]
----
filters:
- type: EnvelopeEncryption                                        # <1>
  config:
    kms: <KMS service name>                                       # <2>
    kmsConfig:                                                    # <3>
      ..:
    selector: <KEK selector service name>                         # <4>
    selectorConfig:                                               # <5>
      ..:
----

> <1> The name of the filter. This must be `EnvelopeEncryption`.

Reviewer 3: I hesitate to use `name` here as its not a something used to identify the instance of the filter. I wonder if it should be the `identifier` for the filter. Or maybe I'm just over thinking it.

<2> The KMS service name.
<3> Object providing configuration understood by KMS provider.
<4> The KEK selector service name.
<5> Object providing configuration understood by key selector.

==== KMS Service configuration

In order to configure the KMS Service, follow these KMS provider specific steps.

include::hashicorp-vault/service.adoc[leveloffset=3]

==== KEK selector configuration

The role of the KEK selector is to map from the topic name to key name.  The filter looks up the resulting
key name in the KMS.

NOTE: If the filter is unable to find the key in the KMS, the filter will pass through the

> records belonging to that topic in the produce request without encrypting them.

Reviewer 3: I still think this represents a bug. As it allows silent failure of the selector policy.

Reviewer 3: We already have the issue https://github.com/kroxylicious/kroxylicious/issues/811

===== Template KEK Selector

The `TemplateKekSelector` maps from topic name to key name.  The template understands the substitution token
`$\{topicName}` which is replaced by the name of the topic.  It can be used to build key names
that include the topic name being encrypted.

Use the `$\{topicName}` is optional. It is possible to pass a literal string.  This will result in all topics being
encrypted using the same key.

[source, yaml]
----
selector: TemplateKekSelector                                 # <1>
selectorConfig:

>   template: "topic.${topicName}"                                # <2>

Reviewer 3: I don't think `topic.${topicName}` is a good example (as it would be a very odd name in the KMS) how about 
```suggestion
  template: "${topicName}_KEK"                                # <2>
```

Author: Good point, but as HashiCorp Vault ACL globbing only supports the wildcard in the last position, let's use a `KEK_` prefix.
